torchtitan/experiments/auto_parallel/parallelize_llama.py

```diff
@@ -33,28 +33,65 @@ def parallelize_llama(
     the model must fit on GPU or CPU memory.
     """
     world_mesh = parallel_dims.world_mesh
-    def input_fn():
-        global_batch_size = job_config.training.global_batch_size
-        if global_batch_size < 0:
-            # This global batch size results in 1 gradient accumulation
-            # step.
-            dp_degree = parallel_dims.dp_replicate * parallel_dims.dp_shard
-            global_batch_size = job_config.training.local_batch_size * dp_degree
-        return (
-            torch.randint(
-                0,
-                # job_config.training.vocab_size,
-                model.vocab_size,
-                (global_batch_size, job_config.training.seq_len),
-                device=torch.device("cuda"),
-            ),
-        )
 
     # TODO make autop work correctly with different combinations of DP, DP+TP, TP, and support DDP / HSDP
     assert parallel_dims.dp_replicate_enabled is False, "DDP not supported yet"
     assert parallel_dims.cp_enabled is False, "CP not supported yet"
-    assert parallel_dims.pp_enabled is False, "PP not supported yet"
 
+    pp_degree = job_config.parallelism.pipeline_parallel_degree
+    local_batch_size = job_config.training.local_batch_size
+    spmd_batch_size = local_batch_size
+    spmd_mesh = world_mesh
+    if parallel_dims.pp_enabled:
+        pp_rank = world_mesh["pp"].get_local_rank()
+        spmd_dims = []
+        if parallel_dims.dp_replicate_enabled:
+            spmd_dims.append("dp_replicate")
+        if parallel_dims.dp_shard_enabled:
+            spmd_dims.append("dp_shard")
+        if parallel_dims.tp_enabled:
+            spmd_dims.append("tp")
+        spmd_mesh = world_mesh[spmd_dims]
+
+        dp_degree = 1
+        for dim in ["dp_replicate", "dp_shard"]:
+            if dim in spmd_mesh.mesh_dim_names:
+                dp_degree *= spmd_mesh[dim].size()
+
+        microbatch_size = job_config.parallelism.pipeline_parallel_microbatch_size
+        n_microbatches = local_batch_size // microbatch_size
+        assert microbatch_size >= 1, f"invalid config {local_batch_size=}, {n_microbatches=}"
+        spmd_batch_size = microbatch_size * dp_degree
+    logger.info(f"Using {spmd_batch_size=}")
+
+    def input_fn():
+        # TODO(whc) - i am not clear what we put this code here for in the first place
+        # if global_batch_size < 0:
+        #     # This global batch size results in 1 gradient accumulation
+        #     # step.
+        #     dp_degree = parallel_dims.dp_replicate * parallel_dims.dp_shard
+        #     global_batch_size = job_config.training.local_batch_size * dp_degree
+        if parallel_dims.pp_enabled and pp_rank > 0:
```

Comment:
What a mess. No action here needed, but it's definitely worth thinking about what the terminal UX state here should be.